compiler-lib/src/semantics.rs

```diff
@@ -0,0 +1,162 @@
+        node: &NodeKind<'a, 'f>,
+    ) -> Result<(), Error> {
+        use self::{ast, NodeKind::*};
+        match node {
+            AST(ast::AST::Program(p)) => {
+                for class in &p.data.classes {
+                    for member in &class.members {
+                        let discr = ast::ClassMemberKindDiscriminants::from(&member.kind);
+                        self.classes_and_members.insert(
+                            (Rc::clone(&class.name), Rc::clone(&member.name), discr),
+                            &member,
+                        );
+                    }
+                }
+                Ok(())
+            }
+            x => {
+                return Err(format_err!(
+                    "unexpected ast node passed to first pass: {:?}",
+                    x
+                ))
+            }
+        }
+    }
+}
+
+type ScopeDefs<T> = HashMap<Symbol, T>;
+
+struct Scope<T> {
+    parent: Option<Rc<Scope<T>>>,
+    definitions: ScopeDefs<T>,
+}
+
+impl<T> Scope<T> {
+    pub fn root() -> Rc<Self> {
+        Rc::new(Scope {
+            parent: None,
+            definitions: ScopeDefs::new(),
+        })
+    }
+    pub fn enter_scope(parent: Rc<Self>) -> Self {
+        Scope {
+            parent: Some(parent),
+            definitions: ScopeDefs::new(),
+        }
+    }
+    pub fn leave(self) -> Option<Rc<Self>> {
+        self.parent
+    }
+    pub fn define(&mut self, sym: &Symbol, val: T) -> Result<(), ()> {
+        use std::collections::hash_map::Entry;
+        let key = Rc::clone(sym);
+        match self.definitions.entry(key) {
+            Entry::Occupied(o) => return Err(()),
+            Entry::Vacant(e) => e.insert(val),
+        };
+        Ok(())
+    }
+}
+
+struct SecondPass;
+
+// TODO proper enum, but Type is definitely necessary
+type VarDef<'a, 'f> = &'a Spanned<'f, ast::Type>;
+struct TyDef; // TODO
+struct MethDef; // TODO
+
+impl SecondPass {
+    pub fn visit<'a, 'f>(
+        ast: &'a ast::AST<'f>,
+        context: &context::Context<'_>,
+        _classes_and_members: &ClassesAndMembers<'a, 'f>,
+    ) -> Result<(), Error> {
+        // TODO java.lang.System.out scope
+        // TODO ontop of that, scope of this package
+        let mut vardefs: Rc<Scope<VarDef<'a, 'f>>> = Scope::root();
+        let mut tydefs: Rc<Scope<TyDef>> = Scope::root();
+        let mut methdefs: Rc<Scope<MethDef>> = Scope::root();
+        Self::do_visit(&NodeKind::from(ast), context, vardefs, tydefs, methdefs)
+    }
+
+    fn do_visit<'a, 'f>(
+        n: &NodeKind<'a, 'f>,
+        context: &context::Context<'_>,
+        vardefs: Rc<Scope<VarDef<'a, 'f>>>,
+        tydefs: Rc<Scope<TyDef>>,
+        methdefs: Rc<Scope<MethDef>>,
+    ) -> Result<(), Error> {
+        use self::NodeKind::*;
+        match n {
+            ClassMember(m) if m.kind.is_method() => {
+                // Assume types have been checked
+                // TODO enforce somewhere that main has String[] arg type
```

Comment:
We wouldn't have to enforce this, if the `MainMethod` had just a `Symbol` and not `ParameterList`...

Reply:
that TODO is legacy. again, we need the variable name for the scope of the method body, so it makes sense to unify main method and normal method for that case.

Reply:
see my comment at line 141 for my solution to that